src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  const clone = { ...state };
+
+  for (const value of actions) {
+    switch (value.type) {
+      case 'addProperties':
+        for (const key in value.extraData) {
+          clone[key] = value.extraData[key];
+        }
+        result.push({ ...clone });
+        break;
+
+      case 'removeProperties':
+        for (const key of value.keysToRemove) {
+          if (key in clone) {
+            delete clone[key];
+          }
+        }
+        result.push({ ...clone });
+        break;
+
+      default:
+        for (const key in clone) {
+          delete clone[key];
+        }
+        result.push({ ...clone });
+    }
+  }
```

Comment:
1. use decomposition
2. throw error in default 
3. move common command out of cases
```suggestion
  for (const { type, .... } of actions) {
    switch (value.type) {
      case 'addProperties':
        for (const key in value.extraData) {
          clone[key] = value.extraData[key];
        }
       
        break;

      case 'removeProperties':
        for (const key of value.keysToRemove) {
          if (key in clone) {
            delete clone[key];
          }
        }
        
        break;

      default:
        // handle all possible cases in case, and throw error in default
    }
    
     result.push({ ...clone });
  }
```

Reply:
If i understood point 3 correctly...